apps/hubble/src/addon/src/trie/merkle_trie.rs

```diff
@@ -0,0 +1,616 @@
+use super::trie_node::TrieNode;
+use crate::{
+    db::{RocksDB, RocksDbTransactionBatch},
+    logger::LOGGER,
+    statsd::statsd,
+    store::{encode_node_metadata_to_js_object, get_merkle_trie, hub_error_to_js_throw, HubError},
+    THREAD_POOL,
+};
+use neon::object::Object as _;
+use neon::{
+    context::ModuleContext,
+    result::NeonResult,
+    types::{buffer::TypedArray as _, JsArray, JsObject},
+};
+use neon::{
+    context::{Context as _, FunctionContext},
+    result::JsResult,
+    types::{Finalize, JsBox, JsBuffer, JsPromise, JsString},
+};
+use slog::{info, o};
+use std::{
+    borrow::Borrow,
+    collections::HashMap,
+    sync::{
+        atomic::{AtomicBool, AtomicU64},
+        Arc, RwLock,
+    },
+};
+
+const TRIE_DBPATH_PREFIX: &str = "trieDb";
+const TRIE_UNLOAD_THRESHOLD: u64 = 10_000;
+
+pub struct NodeMetadata {
+    pub prefix: Vec<u8>,
+    pub num_messages: usize,
+    pub hash: String,
+    pub children: HashMap<u8, NodeMetadata>,
+}
+
+pub struct TrieSnapshot {
+    pub prefix: Vec<u8>,
+    pub excluded_hashes: Vec<String>,
+    pub num_messages: usize,
+}
+
+#[derive(Debug)]
+pub struct MerkleTrie {
+    root: RwLock<Option<TrieNode>>,
+    db: Arc<RocksDB>,
+    logger: slog::Logger,
+    op_count: AtomicU64,
+    db_owned: AtomicBool,
+}
+
+// Implement Finalize so we can pass this struct between JS and Rust
+impl Finalize for MerkleTrie {}
+
+impl MerkleTrie {
+    pub fn new(main_db_path: &str) -> Result<Self, HubError> {
+        let db = Arc::new(RocksDB::new(
```

Comment:
suggestion: I think it could be worth using path joining here

```suggestion
        let path = Path::join(Path::new(main_db_path), Path::new(TRIE_DBPATH_PREFIX))
            .into_os_string()
            .into_string()
            .map_err(|os_str| {
                HubError::validation_failure(format!("error with Merkle Trie path {:?}", os_str))  
            })?;
        let db = Arc::new(RocksDB::new(path.as_str())?);
```